### PR TITLE
feat(jigsaw): Implement the Jigsaw env

### LINKED ISSUE #143: feat: implement full jigsaw environment
Implement a jumanji jigsaw environment.

### PR DESCRIPTION
**Pull request by RuanJohn**
Implements the full Jigsaw environment with actor critic networks. 
closes #143

### CODE REVIEW

File: docs/environments/jigsaw.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+We provide here a Jax JIT-able implementation of a simple _jigsaw_ puzzle. The goal of the agent is to place
+all the jigsaw pieces in the correct locations on an empty 2D puzzle board. Each time an episode resets a
+new puzzle and set of piece is created. Pieces are randomly shuffled and rotated.
+
+## Observation
+The observation given to the agent gives a view of the current state of the puzzle as well as
+all pieces that can be placed.
+
+- `current_board`: jax array (float32) of shape `(num_rows, num_cols)` with values in the range
+    `[1, num_pieces]` (corresponding to the number of each piece). This board will have zeros
```
Comment (sash-a) [inline docs/environments/jigsaw.md:16]: Board will have zeros where no pieces have been placed. So we need to add 0 to action space

```suggestion
    `[0, num_pieces]` (corresponding to the number of each piece). This board will have zeros
```

File: docs/environments/jigsaw.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    where no pieces have been placed and numbers corresponding to each piece where that particular
+    pieces has been paced.
+
+- `pieces`: jax array (float32) of shape `(num_pieces, 3, 3)` of all possible pieces in the
+    current puzzle. These pieces are shuffled and rotated. Pieces will always have shape `(3, 3)`.
+
+- `action_mask`: jax array (bool) of shape `(num_pieces, 4, num_rows-3, num_cols-3)`, representing
```
Comment (sash-a) [inline docs/environments/jigsaw.md:23]: As per our chat offline I think this should be `-2`

```suggestion
- `action_mask`: jax array (bool) of shape `(num_pieces, 4, num_rows-2, num_cols-2)`, representing
```

File: docs/environments/jigsaw.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    which actions are possible given the current state of the board. The first index indicates the
+    number of pieces in a given puzzle. The second index indicates the number of times a piece may be rotated.
+    The third and fourth indices indicate the row and column coordinate of where a piece may be placed respectively.
+    These values will always be `num_rows-3` and `num_cols-3` respectively to make it impossible for an agent to
```
Comment (sash-a) [inline docs/environments/jigsaw.md:27]:
```suggestion
    These values will always be `num_rows-2` and `num_cols-2` respectively to make it impossible for an agent to
```

File: docs/environments/jigsaw.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    place a piece outside the current board.
+
+
+## Action
+The action space is a `MultiDiscreteArray`, specifically a tuple of an index between 0 and `num_pieces`,
+an index between 0 and 4 (since there are 4 possible rotations), an index between 0 and `num_rows-3`
+(the possible row coordinates for placing a piece) and an index between 0 and `num_cols-3`
```
Comment (sash-a) [inline docs/environments/jigsaw.md:34]:
```suggestion
an index between 0 and 4 (since there are 4 possible rotations), an index between 0 and `num_rows-2`
(the possible row coordinates for placing a piece) and an index between 0 and `num_cols-2`
```

**Review by sash-a — COMMENTED** [review]
Just had a look at the docs, will check the rest tomorrow :smile:

File: jumanji/environments/packing/jigsaw/env.py [hunk 5]
```diff
@@ -0,0 +1,523 @@
...
+from jumanji.environments.packing.jigsaw.utils import compute_grid_dim, rotate_piece
+from jumanji.environments.packing.jigsaw.viewer import JigsawViewer
+from jumanji.types import TimeStep, restart, termination, transition
+from jumanji.viewer import Viewer
+
+
+class Jigsaw(Environment[State]):
+
+    """A Jigsaw solving environment with a configurable number of row and column pieces.
+    Here the goal of an agent is to solve a jigsaw puzzle by placing pieces
```
Comment (DriesSmit) [inline jumanji/environments/packing/jigsaw/env.py:40]: @clement-bonnet asked that we just tab lines following the first line.

File: jumanji/environments/packing/jigsaw/env.py [hunk 6]
```diff
@@ -0,0 +1,523 @@
...
+
+    def _check_action_is_legal(
+        self,
+        action: chex.Numeric,
+        current_board: chex.Array,
+        placed_pieces: chex.Array,
+        grid_mask_piece: chex.Array,
+    ) -> bool:
+        """Checks if the action is legal by considering the action mask and the
+           board mask. An action is legal if the action mask is True for that action
```
Comment (DriesSmit) [inline jumanji/environments/packing/jigsaw/env.py:356]: Just tab the lines that are not the first line.

**Review by DriesSmit — COMMENTED** [review]
Thanks @RuanJohn :raised_hands: This looks great :slightly_smiling_face: I will do a more detailed review tomorrow.

File: jumanji/environments/packing/jigsaw/env.py (hunk 5 above)
Comment (RuanJohn) [reply jumanji/environments/packing/jigsaw/env.py]: Done :)

File: jumanji/environments/packing/jigsaw/env.py (hunk 6 above)
Comment (RuanJohn) [reply jumanji/environments/packing/jigsaw/env.py]: Done :)

File: jumanji/environments/packing/jigsaw/conftest.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+def piece_one_correctly_placed(board_with_piece_one_placed: chex.Array) -> chex.Array:
+    """A 2D array of zeros where piece one has been placed correctly."""
+
+    return board_with_piece_one_placed
+
+
+@pytest.fixture()
+def piece_one_partially_placed(board_with_piece_one_placed: chex.Array) -> chex.Array:
+    """A 2D array of zeros where piece one has been placed partially correctly.
+    That is to say that there is overlap between where the piece has been placed and
```
Comment (DriesSmit) [inline jumanji/environments/packing/jigsaw/conftest.py:78]: Can you add the correct tabs throughout the codebase, please :)

File: jumanji/training/configs/config.yaml [hunk 8]
```diff
@@ -1,8 +1,8 @@
 defaults:
     - _self_
-    - env: snake  # [bin_pack, cleaner, connector, cvrp, game_2048, job_shop, knapsack, maze, minesweeper, rubiks_cube, snake, tsp]
+    - env: jigsaw # [bin_pack, cleaner, connector, cvrp, game_2048, jigsaw, job_shop, knapsack, maze, minesweeper, rubiks_cube, snake, tsp]
 
-agent: random  # [random, a2c]
```
Comment (sash-a) [inline jumanji/training/configs/config.yaml:5]: Change this back please

File: jumanji/environments/packing/jigsaw/types.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    step_count: number of steps taken in the environment.
+    key: random key used for board generation.
+    """
+
+    row_nibs_idxs: chex.Array  # (num_sig_rows,)
+    col_nibs_idxs: chex.Array  # (num_sig_cols,)
+    num_pieces: chex.Numeric  # ()
+    solved_board: chex.Array  # (num_rows, num_cols)
+    pieces: chex.Array  # (num_pieces, 3, 3)
+    action_mask: chex.Array  # (num_pieces, num_rotations, num_rows-3, num_cols-3)
```
Comment (sash-a) [inline jumanji/environments/packing/jigsaw/types.py:67]: Are you calculating the action mask from scratch each step? If so you don't need it in the state

File: jumanji/environments/packing/jigsaw/reward.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+            is_valid,
+            lambda: jnp.sum(jnp.equal(state.solved_board, action), dtype=jnp.float32),
+            lambda: jnp.float32(0.0),
+        )
+
+        return reward
+
+
+class SparseReward(RewardFn):
+    """Reward function for the dense reward setting."""
```
Comment (sash-a) [inline jumanji/environments/packing/jigsaw/reward.py:69]: Again give a description of the reward.

```suggestion
    """Reward function for the sparse reward setting."""
```

**Review by sash-a — COMMENTED** [review]
Types and reward :heavy_check_mark:

File: jumanji/environments/packing/jigsaw/env_test.py [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+            num_col_pieces=3,
+            num_row_pieces=3,
+        ),
+    )
+
+
+@pytest.fixture
+def simple_env_board_state_1() -> chex.Array:
+    """The state of the board in the simplified example after 1 correct action."""
+    # fmt: off
```
Comment (djbyrne) [inline jumanji/environments/packing/jigsaw/env_test.py:47]: what is fmt?

File: jumanji/environments/packing/jigsaw/generator.py [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+        """Call method responsible for generating a new state.
+
+        Args:
+            key: jax random key in case stochasticity is used in the instance generation process.
+
+        Returns:
+            A `JigSaw` environment state.
+        """
+
+        raise NotImplementedError
```
Comment (djbyrne) [inline jumanji/environments/packing/jigsaw/generator.py:60]: I don't think this is needed as it is an abstract method

File: jumanji/environments/packing/jigsaw/generator.py [hunk 13]
```diff
@@ -0,0 +1,401 @@
...
+
+        grid = jax.lax.dynamic_update_slice(grid, filled_grid_slice, (0, col - 1))
+
+        return (grid, key), col
+
+    def _select_row_nibs(
+        self, carry: Tuple[chex.Array, chex.PRNGKey], row: int
+    ) -> Tuple[Tuple[chex.Array, chex.PRNGKey], int]:
+        """Creates the nibs for puzzle pieces along rows by randomly selecting
+        a value from the piece above and below the current piece."""
```
Comment (djbyrne) [inline jumanji/environments/packing/jigsaw/generator.py:160]:
```suggestion
        a value from the piece above and below the current piece.
        """
```

File: jumanji/environments/packing/jigsaw/reward.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+import abc
+
+import chex
+import jax
+import jax.numpy as jnp
+
+from jumanji.environments.packing.jigsaw.types import State
+
+
+class RewardFn(abc.ABC):
```
Comment (djbyrne) [inline jumanji/environments/packing/jigsaw/reward.py:24]: I thin a protocol would make more sense here than a base class as it is just the callable? @clement-bonnet

File: jumanji/environments/packing/jigsaw/env.py [hunk 15]
```diff
@@ -0,0 +1,527 @@
...
+        # Unpack and use actions
+        piece_idx, rotation, row_idx, col_idx = action
+
+        chosen_piece = state.pieces[piece_idx]
+
+        # Rotate chosen piece
+        chosen_piece = rotate_piece(chosen_piece, rotation)
+
+        grid_piece = self._expand_piece_to_board(chosen_piece, row_idx, col_idx)
+        grid_mask_piece = self._get_ones_like_expanded_piece(grid_piece=grid_piece)
```
Comment (sash-a) [inline jumanji/environments/packing/jigsaw/env.py:178]: I think this works and gets rid of the extra method

```suggestion
        grid_mask_piece = grid_piece == piece_idx
```

File: jumanji/environments/packing/jigsaw/env.py [hunk 16]
```diff
@@ -0,0 +1,527 @@
...
+
+        action_is_legal = self._check_action_is_legal(
```
Comment (sash-a) [inline jumanji/environments/packing/jigsaw/env.py:180]: Could we rename this to `_is_legal_action`

File: jumanji/environments/packing/jigsaw/env.py [hunk 17]
```diff
@@ -0,0 +1,527 @@
...
+        )
+        grids = batch_expand_piece_to_board(rotated_pieces, rows, cols)
+
+        batch_get_ones_like_expanded_piece = jax.vmap(
+            self._get_ones_like_expanded_piece, in_axes=(0)
+        )
+        grids = batch_get_ones_like_expanded_piece(grids)
+        return grids
+
+    def _make_full_action_mask(
```
Comment (sash-a) [inline jumanji/environments/packing/jigsaw/env.py:468]: I think just `make_action_mask` is fine here

File: jumanji/environments/packing/jigsaw/env.py [hunk 18]
```diff
@@ -0,0 +1,527 @@
...
+        )
+
+        all_possible_pieces = pieces[piece_idxs]
+        rotated_pieces = jax.vmap(rotate_piece, in_axes=(0, 0))(
+            all_possible_pieces, rotations
+        )
+        grids = batch_expand_piece_to_board(rotated_pieces, rows, cols)
+
+        batch_get_ones_like_expanded_piece = jax.vmap(
+            self._get_ones_like_expanded_piece, in_axes=(0)
```
Comment (sash-a) [inline jumanji/environments/packing/jigsaw/env.py:463]: If you delete `_get_ones_like_expanded_piece` then:

```suggestion
            lambda x: x != 0, in_axes=(0)
```

File: jumanji/environments/packing/jigsaw/env.py [hunk 19]
```diff
@@ -0,0 +1,527 @@
...
+        Args:
+            state: current state of the environment.
+
+        Returns:
+            True if the environment is done, False otherwise.
+        """
+
+        done: bool = state.step_count >= state.num_pieces
+
+        return done
```
Comment (sash-a) [inline jumanji/environments/packing/jigsaw/env.py:346]:
```suggestion
        return state.step_count >= state.num_pieces
```

File: jumanji/environments/packing/jigsaw/env.py [hunk 20]
```diff
@@ -0,0 +1,527 @@
...
+        max_col_position = self.num_cols - 2
+
+        return specs.MultiDiscreteArray(
+            num_values=jnp.array(
+                [self.num_pieces, 4, max_row_position, max_col_position]
+            ),
+            name="action",
+        )
+
+    def _check_done(self, state: State) -> bool:
```
Comment (sash-a) [inline jumanji/environments/packing/jigsaw/env.py:333]: Could you rename this to `is_done`?

**Review by sash-a — CHANGES_REQUESTED** [review]
Looks great, thanks Ruan! Really minor style points from my side. 

Only thing I haven't looked at is the networks because I assume those are still being tweaked
